cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/cache.go

```diff
@@ -415,21 +462,58 @@ func (c *ResourceWatcherCache) redisKeyFor(unstructuredObj map[string]interface{
 	// https://redis.io/topics/data-types-intro
 	// Try to stick with a schema. For instance "object-type:id" is a good idea, as in "user:1000".
 	// We will use "helmrepository:ns:repoName"
-	s := fmt.Sprintf("%s:%s:%s", c.config.gvr.Resource, namespace, name)
-	return &s, nil
+	return fmt.Sprintf("%s:%s:%s", c.config.gvr.Resource, namespace, name), nil
 }
 
-// this func is meant to be called to make sure cache client waits
-// for the cache to be fully initialized before attempting read the data out
-// of the cache
-func (c *ResourceWatcherCache) checkInit() error {
-	c.initMutex.Lock()
-	defer c.initMutex.Unlock()
-	if !c.initOk {
-		return status.Errorf(
-			codes.FailedPrecondition,
-			"server cache has not been properly initialized")
-	} else {
-		return nil
+// computing a value for a key maybe expensive, e.g. indexing a repo takes a while,
+// so we will do this in a concurrent fashion to minimize the time window and performance
+// impact of doing so
+// TODO (gfichtenholt) this func should probably return an error, if one occurs so the caller
+// may decide what to do with it. One step at a time.
+func (c ResourceWatcherCache) populateWith(items []unstructured.Unstructured) {
+	// max number of concurrent workers computing cache values at the same time
+	const maxWorkers = 10
+
+	type syncValueJob struct {
+		item map[string]interface{}
+	}
+
+	type syncValueJobResult struct{}
+
+	var wg sync.WaitGroup
+	numWorkers := int(math.Min(float64(len(items)), float64(maxWorkers)))
+	requestChan := make(chan syncValueJob, numWorkers)
+	responseChan := make(chan syncValueJobResult, numWorkers)
+
+	// Process only at most maxWorkers at a time
+	for i := 0; i < numWorkers; i++ {
+		wg.Add(1)
+		go func() {
+			for job := range requestChan {
+				// The following loop will only terminate when the request channel is
+				// closed (and there are no more items)
+				c.onAddOrModify(true, job.item)
+				responseChan <- syncValueJobResult{}
+			}
+			wg.Done()
+		}()
+	}
+	go func() {
+		wg.Wait()
+		close(responseChan)
+	}()
+
+	go func() {
+		for _, item := range items {
+			requestChan <- syncValueJob{item.Object}
+		}
+		close(requestChan)
+	}()
+
+	// Start receiving results
+	// The following loop will only terminate when the response channel is closed, i.e.
+	// after the all the requests have been processed
+	for _ = range responseChan {
+		// TODO (gfichtenholt) maybe have onAddOrModify return an error if fails and log it or return it here
```

Comment:
I could be mis-reading, but I don't *think* you nee the `responseChan` at all. It's being sent empty data (`syncValueJobResult{}`) and you're using it only here to wait until all requests have been processed. You're signalling that by closing the `responseChan` when `wg.Wait()` completes, whereas I think you could just do `wg.Wait()` here at the end of your fn, rather than in the go-func (and hence not need `responseChan` at all? (That's exactly what waitgroups are for).

Reply:
yeah the idea was that later I would add an error to `syncValueJobResult{}` struct in next commit so as to be able to signal an error to the caller if that happens. There is a comment in there to that effect